### PR TITLE
adds new source facebook

### LINKED ISSUE #892: Facebook Certificate Transparency (New Source)
Please describe your feature request:

Add new source Facebook Certificate Transparency (https://developers.facebook.com/tools/ct/search/)

image

### PR DESCRIPTION
**Pull request by tarunKoyalwar**
## Proposed Changes

- adds new source facebook (requires api key)
- closes #892

### CODE REVIEW

**Comment by tarunKoyalwar** [comment]
```console
$ ./subfinder -d hackerone.com -s facebook 

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/

		projectdiscovery.io

[INF] Current subfinder version v2.6.1 (latest)
[INF] Loading provider config from /Users/tarun/.config/subfinder/provider-config.yaml
[INF] Enumerating subdomains for hackerone.com
events.hackerone.com
links.hackerone.com
support.hackerone.com
mta-sts.forwarding.hackerone.com
mta-sts.hackerone.com
api.hackerone.com
gslink.hackerone.com
design.hackerone.com
info.hackerone.com
go.hackerone.com
docs.hackerone.com
mta-sts.managed.hackerone.com
www.hackerone.com
[INF] Found 13 subdomains for hackerone.com in 696 milliseconds 883 microseconds
```

File: v2/pkg/subscraping/sources/facebook/ctlogs_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	response := &response{}
+	if err := json.Unmarshal(bin, response); err != nil {
+		t.Fatal(err)
+	}
+	if len(response.Data) == 0 {
+		t.Fatal("no data found")
+	}
+}
+
+func updateWithEnv(key *string) {
```
Comment (Mzack9999) [inline v2/pkg/subscraping/sources/facebook/ctlogs_test.go:54]: Looks a good candidate for `utils` package!

File: v2/pkg/subscraping/sources/facebook/ctlogs_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/projectdiscovery/utils/generic"
+)
+
+var (
+	fb_API_ID     = "$FB_APP_ID"
+	fb_API_SECRET = "$FB_APP_SECRET"
+)
+
+func TestFacebookSource(t *testing.T) {
+	updateWithEnv(&fb_API_ID)
```
Comment (Mzack9999) [inline v2/pkg/subscraping/sources/facebook/ctlogs_test.go:21]: We can make `updateWithEnv` variadic:

```go
func updateWithEnv(keys ...*string)
```
```

File: v2/pkg/subscraping/sources/facebook/ctlogs.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	urlutil "github.com/projectdiscovery/utils/url"
+)
+
+// source: https://developers.facebook.com/tools/ct
+// api-docs: https://developers.facebook.com/docs/certificate-transparency-api
+// ratelimit: ~20,000 req/hour per appID https://developers.facebook.com/docs/graph-api/overview/rate-limiting/
+
+var (
+	domainsPerPage = "1000"
+	authUrl        = "https://graph.facebook.com/oauth/access_token?client_id=%s&client_secret=%s&grant_type=client_credentials"
```
Comment (Mzack9999) [inline v2/pkg/subscraping/sources/facebook/ctlogs.go:24]: Unrelated and optional: somehow I think that string holding fmt formats should be more explicit or easily identifiable, with naming or maybe defining a new alias type? What do you think of implementing an enriched StringFmt type like:

```
// with naming
var variableNameFmt = "xxx%s"

// or with alias type
// var xxx StringFmt = "xxx%s"
// xxx.Printf(arg[0], arg[1])

type StringFmt string

func (s StringFmt) Printf(args ...interface{}) { ... }
```

**Review by Mzack9999 — APPROVED** [review]
implementation: lgtm!
Todo:
- [ ] Functional testing (currently unable to get an api key)

**Review by ehsandeep — APPROVED** [review]
```console
./subfinder -d hackerone.com -s facebook

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/

		projectdiscovery.io

[INF] Current subfinder version v2.6.1 (latest)
[INF] Loading provider config from /Users/geekboy/.config/subfinder/provider-config.yaml
[INF] Enumerating subdomains for hackerone.com
mta-sts.hackerone.com
design.hackerone.com
events.hackerone.com
info.hackerone.com
mta-sts.managed.hackerone.com
api.hackerone.com
support.hackerone.com
mta-sts.forwarding.hackerone.com
gslink.hackerone.com
go.hackerone.com
links.hackerone.com
docs.hackerone.com
www.hackerone.com
[INF] Found 13 subdomains for hackerone.com in 796 milliseconds 607 microseconds
```

File: v2/pkg/subscraping/sources/facebook/ctlogs_test.go (hunk 1 above)
Comment (tarunKoyalwar) [reply v2/pkg/subscraping/sources/facebook/ctlogs_test.go:54]: https://github.com/projectdiscovery/utils/issues/223

**Pull request merged**